vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go

```diff
@@ -444,6 +444,17 @@ Loop:
 		select {
 		case oomInfo := <-feeder.oomChan:
 			klog.V(3).Infof("OOM detected %+v", oomInfo)
+			// On JVM Heap OOM, we need special handling.
+			// We must also add an RSS peak OOM (with value of memory limit), in order to keep the delta between RSS and JVM Heap the same or higher.
+			// When adding a JVM Heap OOM sample, we add a sample in the next biggest histogram bucket.
+			// Since RSS value is guaranteed to be greater than Heap, we add an oom sample in the RSS histogram as well,
+			// which is guaranteed to be the same increase or larger.
+			if oomInfo.Resource == model.ResourceJVMHeapCommitted {
```

Comment:
I've slightly modified the approach after some testing. The old approach turned out to be wrong.
Before, we were adding in (on container aggregate level) a new RSS histogram sample with 100th percentile value of what is in the histogram, while adding in a JVM Heap Committed histogram sample with the value of the env var (set by autopilot, which includes the autopilot safety margin). 

This led to JVM heap recommendations growing faster than RSS, and incorrect state.
This new approach adds the memory limit to the oomInfo struct (no less invasive way to do it afaict), which is only used on JVMHeapCommitted OOM. When this happens, we just record two OOMs, one for rss and one for jvm heap